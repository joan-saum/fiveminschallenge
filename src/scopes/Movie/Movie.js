// @flow

import React, {useMemo} from 'react';
import {Text, Image, View} from 'react-native';
import {useSelector} from 'react-redux';

import type {Node} from 'react';

import {getMovieById} from '../../redux/selectors';

import styles from './Movie.style';

type RouteProp = {|
  params: {|
    id: number,
  |},
|};

type Props = {|
  route: RouteProp,
|};

const Movie = ({route}: Props): Node => {
  const {id} = route.params;
  const movie = useSelector(state => getMovieById(state, id));

  const source = useMemo(
    () => ({
      uri: movie ? movie.image : '',
    }),
    [movie],
  );

  if (!movie) {
    return null;
  }

  return (
    <View>
      <Image source={source} style={styles.image} />
      <Text style={styles.title}>{movie.title}</Text>
      <Text style={styles.overview}>{movie.overview}</Text>
    </View>
  );
};

export default Movie;
