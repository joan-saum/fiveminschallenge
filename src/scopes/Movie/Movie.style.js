import {StyleSheet} from 'react-native';

const style = StyleSheet.create({
  image: {
    width: '100%',
    height: 300,
    marginBottom: 10,
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
    marginBottom: 5,
    marginLeft: 15,
    marginRight: 15,
  },
  overview: {
    marginLeft: 10,
    marginRight: 10,
  },
});

export default style;
