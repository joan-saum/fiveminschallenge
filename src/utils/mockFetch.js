// @flow

import fetch from 'jest-fetch-mock';
import type {Store, Reducer, StoreEnhancer} from 'redux';
import {createStore, applyMiddleware, compose} from 'redux';
import {middleware as reduxPackMiddleware} from 'redux-pack';

import type {Action, Dispatch} from '../redux/types';

export const mockApiSuccess = (response: any) => {
  fetch.mockResponseOnce(JSON.stringify(response), {
    headers: {
      'Content-Type': 'application/json',
    },
  });
};

export const mockApiFailure = (response: any) => {
  fetch.mockResponseOnce(JSON.stringify(response), {
    status: 555,
    headers: {
      'Content-Type': 'application/json',
    },
  });
};

export const mockStore = <S>(
  reducer: Reducer<S, Action>,
  preloadedState?: S,
): Store<S, Action, Dispatch> => {
  const middlewares: StoreEnhancer<S, Action, Dispatch> = compose(
    applyMiddleware(reduxPackMiddleware),
  );

  return createStore(reducer, preloadedState, middlewares);
};

export const configureTests = () => {
  global.fetch = fetch;
};
