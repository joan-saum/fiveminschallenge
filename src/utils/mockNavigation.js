// @flow

import type {NavigationProps} from '../types/navigation';

const navigation: NavigationProps = {
  setParams: jest.fn().mockName('setParams'),
  getParam: jest.fn().mockName('getParam'),
  navigate: jest.fn().mockName('navigate'),
  dangerouslyGetParent: jest.fn().mockName('dangerouslyGetParent'),
  goBack: jest.fn().mockName('goBack'),
  addListener: jest.fn().mockName('addListener'),
  isFocused: jest.fn().mockName('isFocused'),
  push: jest.fn().mockName('push'),
  pop: jest.fn().mockName('pop'),
  popToTop: jest.fn().mockName('popToTop'),
  dismiss: jest.fn().mockName('dismiss'),
  state: {},
};

export default navigation;
