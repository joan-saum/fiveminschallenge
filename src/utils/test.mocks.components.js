// @flow

jest.mock('../components/InputSearch/InputSearch');
jest.mock('../components/MovieItem/MovieItem');
