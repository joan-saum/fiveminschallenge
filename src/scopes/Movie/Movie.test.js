// @flow

import * as React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import {configureTests} from '../../utils/mockFetch';

import Movie from './Movie';

configureTests();

const mockStore = configureStore([]);

jest.unmock('./Movie');

const defaultProps = {
  route: {
    params: {
      id: 1,
    },
  },
};

describe('Movie', () => {
  it('should render a movie', () => {
    const item = {id: 1, overview: 'overview', title: 'title', image: 'image'};
    const store = mockStore({
      items: [item],
    });
    const tree = renderer.create(
      <Provider store={store}>
        <Movie {...defaultProps} />
      </Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('should render nothing when the movie is not found', () => {
    const store = mockStore({
      items: [],
    });
    const tree = renderer.create(
      <Provider store={store}>
        <Movie {...defaultProps} />
      </Provider>,
    );

    expect(tree).toMatchInlineSnapshot('null');
  });
});
