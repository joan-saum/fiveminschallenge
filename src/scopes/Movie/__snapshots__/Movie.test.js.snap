// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie should render a movie 1`] = `
<View>
  <Image
    source={
      Object {
        "uri": "image",
      }
    }
    style={
      Object {
        "height": 300,
        "marginBottom": 10,
        "width": "100%",
      }
    }
  />
  <Text
    style={
      Object {
        "fontSize": 20,
        "fontWeight": "bold",
        "marginBottom": 5,
        "marginLeft": 15,
        "marginRight": 15,
      }
    }
  >
    title
  </Text>
  <Text
    style={
      Object {
        "marginLeft": 10,
        "marginRight": 10,
      }
    }
  >
    overview
  </Text>
</View>
`;
